src/main/java/com/pnu/dev/shpaltaif/service/PostServiceImpl.java

```diff
@@ -0,0 +1,126 @@
+package com.pnu.dev.shpaltaif.service;
+
+import com.pnu.dev.shpaltaif.domain.Category;
+import com.pnu.dev.shpaltaif.domain.Post;
+import com.pnu.dev.shpaltaif.domain.User;
+import com.pnu.dev.shpaltaif.domain.UserRole;
+import com.pnu.dev.shpaltaif.dto.PostDto;
+import com.pnu.dev.shpaltaif.dto.PostFiltersDto;
+import com.pnu.dev.shpaltaif.exception.ServiceAdminException;
+import com.pnu.dev.shpaltaif.exception.ServiceException;
+import com.pnu.dev.shpaltaif.repository.CategoryRepository;
+import com.pnu.dev.shpaltaif.repository.PostRepository;
+import com.pnu.dev.shpaltaif.repository.specification.PostSpecificationBuilder;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Service;
+
+import java.time.LocalDateTime;
+
+@Service
+public class PostServiceImpl implements PostService {
+
+    private final PostRepository postRepository;
+
+    private final CategoryRepository categoryRepository;
+
+    private final PostSpecificationBuilder postSpecificationBuilder;
+
+    @Autowired
+    public PostServiceImpl(PostRepository postRepository, CategoryRepository categoryRepository, PostSpecificationBuilder postSpecificationBuilder) {
+        this.postRepository = postRepository;
+        this.categoryRepository = categoryRepository;
+        this.postSpecificationBuilder = postSpecificationBuilder;
+    }
+
+    @Override
+    public Page<Post> findAll(User user, PostFiltersDto filtersDto, Pageable pageable) {
+
+        Specification<Post> specification = postSpecificationBuilder.buildPostSpecification(user, filtersDto);
+        return postRepository.findAll(specification, pageable);
+    }
+
+    @Override
+    public Post findActiveById(Long id) {
+        return postRepository.findByIdAndActiveTrue(id)
+                .orElseThrow(() -> new ServiceException("Пост не знайдено"));
+    }
+
+    @Override
+    public Post findById(User user, Long id) {
+
+        Post post = postRepository.findById(id)
+                .orElseThrow(() -> new ServiceAdminException("Пост не знайдено"));
+
+        if (user.getRole().equals(UserRole.ROLE_ADMIN)) {
```

Comment:
Don't like the idea that half of the application security related to posts are handled by this method.
It doesn't look supportable for me.
But as long as logic is relatively simple and obvious, I suggest to create separate private method in this class with name "findByIdAllowingWriteAccessPermissions" for now, just to comunicate the purpose of this method (and maybe we redesign one day if we will see a lot of such problems).

Reply:
Actually, I would rather split this logic.
The problem here is that if user has role ADMIN the post will be returned, but it's not really what we expect from application point of view, because admin cannot edit posts, for example, because he doesn't have writer account. Spring security restricts such access, but it's confusing anyway.

so it's better to have the following methods:

```
private Post findById(Long id) {
        return postRepository.findById(id)
                .orElseThrow(() -> new ServiceAdminException("Пост не знайдено"));
}
```

```
private void checkWriterAccessPermissions(Post post, User user) {
if (!StringUtils.equals(post.getAuthorPublicAccount().getId(), user.getPublicAccount().getId())) {
            throw new ServiceAdminException("Ви не маєте доступ до цього поста");
        }
}
```

private void checkUserAccessPermissions(Post post, User user) {
     if (user.getRole() == UserRole.ROLE_ADMIN) {
     return;
     }
     checkWriterAccessPermissions(post, user);
}

And then use this methods in each place where it's required. Actually, it's not a code duplication, because current logic do not reflect correct behaviour.